### ISSUE TITLE
api auth errors are now different

### ISSUE THREAD

**Issue by bf4** [minimul/qbo_api#39]
As evidenced by

```
NoMethodError: undefined method `[]' for nil:NilClass"
["/app/vendor/bundle/ruby/2.3.0/gems/qbo_api-1.5.3/lib/qbo_api/raise_http_exception.rb:55:in `parse_json'","/app/vendor/bundle/ruby/2.3.0/gems/qbo_api-1.5.3/lib/qbo_api/raise_http_exception.rb:47:in `error_body'","/app/vendor/bundle/ruby/2.3.0/gems/qbo_api-1.5.3/lib/qbo_api/raise_http_exception.rb:41:in `error_message'","/app/vendor/bundle/ruby/2.3.0/gems/qbo_api-1.5.3/lib/qbo_api/raise_http_exception.rb:15:in `block in call'","/app/vendor/bundle/ruby/2.3.0/gems/faraday-0.13.1/lib/faraday/response.rb:61:in `on_complete'","/app/vendor/bundle/ruby/2.3.0/gems/qbo_api-1.5.3/lib/qbo_api/raise_http_exception.rb:9:in `call'
]
```

Notice that the element names change from Fault -> fault etc. Also, technically, a request with `Accept: application/json` should return a 415 rather than `Content-Type: text/xml`.

```diff
 - request:
      method: get
    uri: https://sandbox-quickbooks.api.intuit.com/v3/company/XXXXX/query?query=SELECT%20*%20FROM%20Account%20MAXRESULTS%201%20STARTPOSITION%201
     headers:
       Content-Type:
       - application/json;charset=UTF-8
       Accept:
       - application/json
   response:
     headers:
       Content-Type:
-      - text/xml
+      - application/json
     body:
        encoding: UTF-8
-      string: |
-        <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
-        <IntuitResponse time="2017-10-30T07:25:42.838-07:00" xmlns="http://schema.intuit.com/finance/v3">
-            <Fault type="AUTHENTICATION">
-                <Error code="3200">
-                    <Message>message=AuthenticationFailed; errorCode=003200; statusCode=401</Message>
-                    <Detail></Detail>
-                </Error>
-            </Fault>
-        </IntuitResponse>
+      string: '{"warnings":null,"intuitObject":null,"fault":{"error":[{"message":"message=AuthenticationFailed;
+        errorCode=003200; statusCode=401","detail":"","code":"3200","element":null}],"type":"AUTHENTICATION"},"report":null,"queryResponse":null,"batchItemResponse":[],"attachableResponse":[],"syncErrorResponse":null,"requestId":null,"time":1512143684083,"status":null,"cdcresponse":[]}'
```

I've hacked around it by patching the middleware

```ruby
require 'qbo_api/raise_http_exception'
module QboApiInconsistentJsonResponseHack
  # Expected Structure:
  # {"Fault":{"Error":[{"Message":"message=AuthenticationFailed; errorCode=003200; statusCode=401",
  #                    "Detail":"",
  #                     "code":"3200","element":""}],
  #                     "type":"AUTHENTICATION"},"time":"2017-11-16T13:48:17.050-08:00"}'
  # fails to be parsed when now returned with different keys
  # {"warnings":null,"intuitObject":null,
  #  "fault":{"error":[{"message":"message=AuthenticationFailed; errorCode=003200; statusCode=401",
  #                     "detail":"","code":"3200","element":null}],
  #                   "type":"AUTHENTICATION"},"report":null,"queryResponse":null,"batchItemResponse":[],"attachableResponse":[],"syncErrorResponse":null,"requestId":null,"time":1512105735900,"status":null,"cdcresponse":[]}
  # usually when was previously XML
  #   <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
  #   <IntuitResponse time="2017-10-30T07:25:42.838-07:00" xmlns="http://schema.intuit.com/finance/v3">
  #       <Fault type="AUTHENTICATION">
  #           <Error code="3200">
  #               <Message>message=AuthenticationFailed; errorCode=003200; statusCode=401</Message>
  #               <Detail></Detail>
  #           </Error>
  #       </Fault>
  #   </IntuitResponse>
  def parse_json(body)
    return super unless body =~ /"fault"/
    res = JSON.parse(body)
    # keys different from expected
    raise unless res.key?("fault")
    fault = res["fault"]
    raise unless (["error", "type"] - fault.keys).empty?
    errors = fault["error"]
    errors.map { |error|
      raise unless (["code", "message", "detail"] - error.keys).empty?
      {
        fault_type: fault["type"],
        error_code: error['code'],
        error_message: error['message'],
        error_detail: error['detail']
      }
    }
  end
end
FaradayMiddleware::RaiseHttpException.prepend(QboApiInconsistentJsonResponseHack)
```

**Comment by ntippie** [comment]
Nicely documented. I just put in a pull request that will fix this behavior.

I think we're seeing the v4 API leaking through the cracks a bit here. It seems that they are moving to a consistent letter case for all keys.

**Comment by minimul** [comment]
Thanks  @bf4  for reporting. Closing #40 . Released in 1.5.4

**Comment by bf4** [comment]
Thanks @ntippie @minimul 

Meanwhile, you fixed it and Intuit support asked why I was getting a 401.
